File: arrays/sortedSquaredArray.cpp
/*
 Given a sorted array of ints
 return sorted array of squares of ints.
 input:[-5, -4, -3, -2, -1]
 output:[1, 4, 9, 16, 25]
 */
#include <vector>
using namespace std;

//O(nlogn) time and O(n) space
vector<int> sortedSquaredArray1(vector<int> array) {
	vector<int> res;
	for(auto a:array){
	  res.push_back(a*a);	
	}
	sort(res.begin(),res.end());
  return res;
}

//O(n) time and O(n) space
vector<int> sortedSquaredArray(vector<int> array) {
	vector<int> res(array.size(),0);
	int l=0, r=array.size()-1;
	for(int i=array.size()-1; i>=0; i--){
		if(abs(array[l])>abs(array[r])){
			res[i]=array[l]*array[l];
			l++;
		}else{
			res[i]=array[r]*array[r];
		  r--;	
		}
	}
	return res;
}

File: arrays/sortedSquaredArray.go
package main
import "sort"

//O(n) time and O(n) space
func SortedSquaredArray1(array []int) []int {
    res := make([]int, len(array))
    i:=len(res)-1
	  l := 0
    r := len(array)-1
    for i>=0 {
        left := array[l]*array[l]
        right := array[r]*array[r]
        if left > right {
            res[i]=left
            l++
        } else {
            res[i]=right
            r--
        }
        i--
    }
	return res
}

//O(nlogn) time and O(1) space
func SortedSquaredArray2(array []int) []int {
    for i, num := range array {
        array[i] = num * num
    }

    sort.Ints(array)
	  return array
}
